content/blog/2024-10-21-testing-the-limits/index.md

```diff
@@ -0,0 +1,243 @@
+```
+
+Test breakdown:
+- `-n` - this will run 1,000,000 operations using the commands in -t
+- `-t` - will run the set and get tests
+- `-P` - this specifies that we would like the tests to use 16 pipelines (send 16 operations per request). 
+- `-q` - silences the output to show only the final results
+- `-a` - use the specified password
+- `-h` - run the test against the specified host
+- `--threads` - how many threads to generate test data from
+
+Honestly, I was asonished by the first set of results I got. Sure, I expected Valkey to be fast but this speed from a single board computer?? ON A SINGLE THREAD?? It's amazing. 
+
+```bash
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad --threads 5 -h 10.0.1.136
+SET: 173040.33 requests per second, p50=4.503 msec                    
+GET: 307031.00 requests per second, p50=2.455 msec
+```
+
+Between the two tests we averaged 240,000 requests per second.  
+
+## Raising the CPU clock speed
+
+Since Valkey is a single threaded application it makes sense that higher clock speeds would lead to more performance. I don't expect most people will overclock their servers in production. Different servers may be available with different CPU clock speeds. 
+
+**Note:** Clock speeds generally are only comparable between CPU's with a similar architecture. For example, you could reasonably compare clock speeds between an 12th generation Intel i5 and a 12th generation Intel i7. If the 12th gen i7 had a max clock speed of 5Ghz that doesn't necessarily mean it will be slower than a AMD Ryzen 9 9900X clocked at 5.6Ghz. 
+
+If you're following along on a Pi of your own I've outlined the steps to overclock your CM4 below. Otherwise you can skip to the results section below.
+
+1. Open the below file in a text editor like nano.
+    - `sudo nano /boot/firmware/config.txt`
+2. At the end of the file add this section below the `[all]` tag.
+    ```
+    [all]
+    over_voltage=8
+    arm_freq=2200
+    ```
+3. Save with cmd + x, y, enter 
+4. Restart the Pi and log back in `sudo restart now`
+
+We've just increased the speed of the Pi by 47% by raising the clock speed from 1.5Ghz to 2.2Ghz. Now lets re-run our test and see how things look!
+
+```bash 
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad --threads 5 -h 10.0.1.136
+SET: 394368.41 requests per second, p50=1.223 msec                    
+GET: 438058.53 requests per second, p50=1.135 msec 
+```
+
+We're up to 416,000 requests per second (reminder this is the average between the two operations). The mathmaticians out there might notice that this speed up is a lot more than the expected 47% increase. It's a 73% increase in requests per second. What's happening?!
+
+I actually had to take this one back to the Valkey team to figure out and we've yet to figure out exactly what is happening here. 
```

Comment:
figure out is said twice, rephrase a bit for clarity

Reply:
I removed the line because I cover this in more detail later.